kale/codegen/generate_code.py

```diff
@@ -3,159 +3,219 @@
+    to /marshal.
+
+    Args:
+        wd: current working directory. Can be None
+        volumes: volumes dictionary
+        nb_path: path to the notebook file
+
+    Returns: (dict): a dict composed of
+        - 'marshal_volume' (bool): True if we use a custom marshal volume
+        - 'marshal_path' (str): path to the volume, if `marshal_volume` is True
+    """
     marshal_volume = True
     marshal_path = "/marshal"
     # Check if the workspace directory is under a mounted volume.
     # If so, marshal data into a folder in that volume,
     # otherwise create a new volume and mount it at /marshal
-    wd = metadata.get('abs_working_dir', None)
     if wd:
         wd = os.path.realpath(wd)
-        # get the volumes for which the working directory is a subpath of the mount point
+        # get the volumes for which the working directory is a subpath of
+        # the mount point
         vols = list(filter(lambda x: wd.startswith(x['mount_point']), volumes))
         # if we found any, then set marshal directory inside working directory
-        if len(vols) >= 1:
+        if len(vols) > 0:
             marshal_volume = False
             marshal_dir = f".{os.path.basename(nb_path)}.kale.marshal.dir"
             marshal_path = os.path.join(wd, marshal_dir)
+    return {
+        'marshal_volume': marshal_volume,
+        'marshal_path': marshal_path
+    }
+
+
+def get_args(pipeline_parameters):
+    """Generate pipeline and function parameter.
+
+    The generated strings will be passed to the rendering template.
 
-    pipeline_args_names = list(pipeline_parameters.keys())
+    Args:
+        pipeline_parameters (dict): pipeline parameters as
+        {<name>:(<type>,<value>)}
+
+    Returns (dict): a dict composed of:
+        - 'pipeline_args_names': pipeline argument names as list
+        - 'pipeline_args': pipeline arguments as a comma separated string
+        - 'function_args': function arguments as a comma separated string
+    """
+    pipeline_args_names = ', '.join(list(pipeline_parameters.keys()))
     # wrap in quotes every parameter - required by kfp
     pipeline_args = ', '.join([f"{arg}='{pipeline_parameters[arg][1]}'"
                                for arg in pipeline_parameters])
-    # arguments are actually the pipeline arguments. Since we don't know precisely in which pipeline
-    # steps they are needed we just pass them to every one. The assumption is that these variables
-    # were treated as constants notebook-wise.
-    function_args = ', '.join([f"{arg}: {pipeline_parameters[arg][0]}" for arg in pipeline_parameters])
+    # Arguments are the pipeline arguments. Since we don't know precisely in
+    # what pipeline steps they are needed, we just pass them to every one.
+    # We assume there variables were not re-assigned throughout the notebook
+    function_args = ', '.join([f"{arg}: {pipeline_parameters[arg][0]}"
+                               for arg in pipeline_parameters])
```

Comment:
Python3.5 compat
```python
function_args = ', '.join(["{}: {}".format(arg,
                                           pipeline_parameters[arg][0])
                           for arg in pipeline_parameters])
```

Reply:
I will leave all PY3.5 `format` refactoring to #54